polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+package polynote.server
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import cats.Monad
+import cats.effect.concurrent.{Deferred, Ref, Semaphore}
+import cats.effect.{Concurrent, ContextShift, IO}
+import cats.syntax.functor._
+import cats.syntax.flatMap._
+import fs2.Stream
+import fs2.concurrent.{Enqueue, Queue, SignallingRef, Topic}
+import polynote.kernel._
+import polynote.kernel.util.{Publish, ReadySignal, WindowBuffer}
+import polynote.messages._
+import polynote.server.IOSharedNotebook.{GlobalVersion, SubscriberId}
+import polynote.util.VersionBuffer
+
+/**
+  * SharedNotebook is responsible for managing concurrent access and updates to a notebook. It's the central authority
+  * for the canonical serialization of edits to the notebook. Eventually, it should resolve conflicts in concurrent edits
+  * and broadcast changes to subscribers.
+  *
+  * Here's the current idea for how this would work (though it's not completely implemented):
+  *
+  * The notebook has a "global" (server) version number, and each client has a "local" (client) version number. When
+  * a client sends an edit to the server, it includes the latest global version it knows about, and its local version,
+  * which it then increments. When the server broadcasts an edit, it sends a message to each subscriber containing
+  * the global version of the edit, and the local version to which it applies for that client. Then, the client can
+  * transform the edit (if necessary) to its current local version, and it will be in a consistent state with the
+  * global version.
+  *
+  * Similarly, the server can do part of the transformation work – when it receives an edit that acts upon global version
+  * X, but it knows that client A is currently on at least version X + 3, it can transform the edit to act upon
+  * version X + 3 and send that transformed edit to client A (which will also apply any necessary transformations locally).
+  *
+  * Thus, the SharedNotebook must track its global version, the current notebook, and low watermark of the highest global version
+  * that all clients have acknowledged. It must also keep a history buffer of all edits between the low watermark and
+  * the current global version. Clients can periodically send a status message to the server indicating the current-known
+  * global version and the current local version, in order to allow the server to discard edit history.
+  *
+  * Clients must track the currently known global version, which updates whenever the client receives a foreign edit,
+  * and the highest local version which has been acknowledged by the server (either by sending an edit based on it,
+  * or through an acknowledgement message). It has to keep its local edit history between that known-acknowledged
+  * version and its current version, to allow rebasing new foreign edits.
+  *
+  * It sounds a lot more complicated than it really is.
```

Comment:
I guess it could be viewed as a kind of a variation of a vector clock (client-server vs. fully distributed) where SharedNotebook keeps the (master) vector clock of `[V(self), min(V(clients))]`, and each client keeps a vector clock of `[V(self), V(latest_global)]`?

Reply:
And everyone is required to keep their own set of transformations between `V(self)` and `V(latest_global)`

Reply:
Yes, kind of, except that the clients don't keep "clocks" for the other clients. I'd be all about a smarter EC/synchronization scheme, but this was the first thing that came to mind and it seemed reasonably straightforward (until I tried to articulate it in a comment 😬 )